lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -387,94 +408,70 @@ def __init__(self, charm: "OpenSearchBaseCharm", relation_name: str = PeerCluste
+            charm.on[PeerClusterRelationName].relation_departed,
+        ]:
+            # We need to keep track of the peer-cluster relation
+            # A unit-level secret will not trigger secret changes
+
+            # I've discussed this offline with @wallyworld and the main idea
+            # is that the unit is already aware of the secret change, why triggering
+            # a new hook in this case?
+            # Now, opensearch_backups.py was originally using opensearch_secrets.py to
+            # update its the s3-credentials secret and inform this class via "manual_update"
+
+            # Listening to the peer cluster relation is another alternative:
+            # Effectively it will call the common method that both _on_secret_changed and
+            # _on_peer_cluster_relation_event uses to update the keystore.
+            self.framework.observe(event, self._on_peer_cluster_relation_event)
+
     @override
-    def manual_update(self, event: EventBase) -> None:
-        """Processes the non-orchestrator cluster events."""
-        if not self.charm.plugin_manager.check_plugin_manager_ready():
-            logger.warning("s3-changed: cluster not ready yet")
+    def _on_secret_changed(self, event: EventBase) -> None:
+        """Clean secret from the plugin cache."""
+        secret = event.secret
+
+        if not event.secret.label:
+            logger.info("Secret %s has no label, ignoring it.", event.secret.id)
             return
 
-        if not (s3_creds := self.charm.secrets.get_object(Scope.APP, "s3-creds")):
+        if S3_CREDENTIALS not in event.secret.label:
+            logger.debug("Secret %s is not s3-credentials, ignoring it.", event.secret.id)
             return
+        secret.get_content(refresh=True)
 
-        s3_creds = S3RelDataCredentials.from_dict(s3_creds)
+        self._on_peer_cluster_relation_event(event)
 
-        # https://github.com/canonical/opensearch-operator/issues/252
-        # We need the repository-s3 to support two main relations: s3 OR peer-cluster
-        # Meanwhile, create the plugin manually and apply it
-        try:
-            plugin = OpenSearchPluginConfig(
-                secret_entries_to_del=[
-                    "s3.client.default.access_key",
-                    "s3.client.default.secret_key",
-                ],
-            )
-            self.charm.plugin_manager.apply_config(plugin)
-        except OpenSearchError as e:
-            logger.warning(
-                f"s3-changed: failed disabling with {str(e)}\n"
-                "repository-s3 maybe it was not enabled yet"
-            )
-        # It must be able to enable the plugin
+    def _on_peer_cluster_relation_event(self, event: EventBase) -> None:
+        """Processes the peer-cluster relation events."""
+        if not self.charm.secrets.get_object(Scope.APP, S3_CREDENTIALS):
+            logger.warning(f"Secret {S3_CREDENTIALS} found but missing s3-credentials set.")
```

Comment:
> This will not evaluate to true the first invocation of this hook.

Why not? Can you provide more details why this `if` will not be true for every `peer-cluster-changed` call?

> Given that this secret is only effectively stored at the end of this hook by the leader

Why?

> In which case, why listen to it altogether?

Simply put, we are only interested in the `secret-changed`. The way we manage secrets is by receiving these secrets via peer-cluster relation and then setting them in the app-level. Now, that triggers secret-changed or secret-removed in some units.

As my comment previously says: unit-level secrets do not trigger secret-changed, nor app-level secrets trigger secret-changed in the app leader. Therefore, the only real way for us to load the secrets coming from the orchestrator *on every unit* is to listen to the peer cluster.

The secret-changed truly only exists to call the `secret.get_content(refresh=True)` and get the secret content updated.